content/blog/2025-05-13-imperative-quantifier-elimination.md

```diff
@@ -0,0 +1,178 @@
+|   14   |   47,308 (2,600)    |           282 (3)                 |
+|   15   |   435 (8)           |           307 (7)                 |
+|   16   |   217,818 (1,517)   |           278 (5)                 |
+|   17   |   838 (16)          |           278 (3)                 |
+|   18   |   311 (6)           |           304 (6)                 |
+|   19   |   55,716 (200)      |           308 (3)                 |
+|   20   |   37,889 (593)      |           370 (9)                 |
+
+These numbers suggest that the quantifier elimination pass tends to have a slight positive improvement on performance and occasionally causes a massive improvement. With a little care we can identify fairly precisely when huge performance improvement spikes occur (and without having to parse the abomination that is randomly generated program number 14). Consider the following quantifier expression:
+
+```Python
+any(x > y for x in range(z))
+```
+we will repeat the experiment above with our sole input `z` ranging from `0` to `5000` and `y` set either to `0`, `2500`, or `5000`, all times reported in seconds:
+
+|     y     | Run Time (Original) | Run Time (Quantifiers Eliminated) |
+|-----------|---------------------|-----------------------------------|
+|     0     |       0.2 (0.01)    |            0.2 (0.02)             |
+|  500 000  |       2.1 (0.03)    |            0.2 (0.00)             |
+| 1 000 000 |       2.6 (0.18)    |            0.2 (0.00)             |
+
+Unsurprisingly, when a quantifier expression terminates early there is not much room for improvement. The real benefits of quantifier elimination come when it can eliminate long-running `any` and `all` calls.
+
+In fact, randomly generated program 16 is pretty much the worst case scenario from this perspective:
+
+```Python
+any(n != n for i in range(n, -19*n))
+```
+
+The pass correctly transforms this to:
+
+```Python
+False
+```
+
+I think it is clear why those two expressions would have very different runtimes for, say, `n=-10,000`. I also hope it is clear that the huge time savings in the first table are mostly the result of the input programs being randomly generated and silly.
+
+### Utility
+
+One final measure of a compiler pass is how applicable it is to code that people actually write. To measure that, I gathered a bunch of open source Python code. I didn't gather it in any principled way, I just grabbed a bunch from the [GitHub trending page](https://github.com/trending/python?since=daily) and from [this collection of Python projects](https://github.com/vinta/awesome-python). 
+
+According to [cloc](https://github.com/AlDanial/cloc), I searched ~1.68 million lines of Python code and found 0 lines that I could apply my quantifier elimination pass to. Possibly that number could be improved by expanding the pass (by re-including the $P_m$s, for example) but I suspect that people just don't write many Presburger Arithmetic quantifier expressions in their Python code.
```

Comment:
Not exactly a shocking result, but too bad nonetheless! I had a fun time using a naive text search query to look for other possible candidates: https://github.com/search?q=%2Fany.*%3C.*for+i+in+range%2F+language%3APython&type=code

And yeah, slim pickings. Here are a few examples that are kinda close, maybe, although I didn't think about it very hard:
https://github.com/arin17bishwa/myCP_sols/blob/1e452b4d192b6df9f897602a1341b194aff15079/CF/1363A.py#L73
https://github.com/CalcoDev/aoc-2024/blob/a3ccbc85bf86b98424a76428605e59747a57611c/day2.py#L7

Reply:
Text search is genius! My search was limited to stuff I could actually apply my pass to so it definitely missed out on some things that were possible but not yet supported. The second example you found could definitely be supported with a little (or maybe a lot of) work.